### PR TITLE
Allow root action rewrites

### PR DESCRIPTION
**Pull request by PascalSenn**
# Explanation 
Since https://github.com/open-telemetry/opentelemetry-dotnet/pull/5026 the activity name of the asp net core span is set on activity end. This means that any change that was done to the name of this span will be overwritten. Many monitoring solutions ( elastic, sentry, datadog et al) use the name of the root span to group similar spans. A tool to control this was to rename the root activity, which is the asp net core activity if you have it enabled. As #5026 overwrites any changes to the root activity on transaction end, this control is gone. 

## Example

The path (or route pattern) for a GraphQL request is always the same. For example `/graphql`. The open telemetry integration of HotChocolate rewrites the name of the root span to the operation that you are execution. Specficially into something like `query FindUserById { userById }`, `mutation updateUserName { updateUserName }` ... . As of #5026, all graphql operations are reported as `POST /graphql/{**slug}` which makes it impossible to properly trace different operations. 

The specific treatment of GRPC in the HttpInListener shows that it would have the same issue, it just happens to be included by default.  

## Why the hooks is not enough
There is a hook which you can use to change the properties of the span. Frameworks like HotChocolate (or others) that want to overwrite the naming of the AspNetCore trace could just use this hook. But this is not a good solution:

1. It requires all frameworks to have a dependency on `OpenTelemetry.Instrumentation.AspNetCore`.
1. It would require the user to add additional configuration to the `AddAspNetCoreInstrumentation`  registration, which leads to confusion. 

## Changes

Before overwriting the activity name on activity end, the `DisplayName` is compared to the one that is set on activity start. If the values are not the same, the user ( or a library), is interested in renaming the name of the span. In this case the name of the span in not modified on activity end. 

## Merge requirement checklist

* [X] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [X] Unit tests added/updated
* [X] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [X] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by vishweshbankwar** [comment]
@PascalSenn Do you mind elaborating bit more on

> The specific treatment of GRPC in the HttpInListener shows that it would have the same issue, it just happens to be included by default.

**Comment by PascalSenn** [comment]
@vishweshbankwar  

The DisplayName of the activity is set on activity end here:
https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs#L244

After this code, there is a special treatment for GRPC:
https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs#L251-L254

Which then overwrites the activity name to the RPC spectific semantic convention:
https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs#L351-L354

If GRPC would not be handled in `OpenTelemetry.Instrumentation.AspNetCore` there would be no way for a GRPC component further down the middleware pipeline to set correct root span name for the RPC semantic convention unless you would register a specific  enricher with`EnrichWithHttpResponse` on the AspNetCore instrumentation

**Comment by vishweshbankwar** [comment]
> @vishweshbankwar
> 
> The DisplayName of the activity is set on activity end here: https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs#L244
> 
> After this code, there is a special treatment for GRPC: https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs#L251-L254
> 
> Which then overwrites the activity name to the RPC spectific semantic convention: https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs#L351-L354
> 
> If GRPC would not be handled in `OpenTelemetry.Instrumentation.AspNetCore` there would be no way for a GRPC component further down the middleware pipeline to set correct root span name for the RPC semantic convention unless you would register a specific enricher with`EnrichWithHttpResponse` on the AspNetCore instrumentation

@PascalSenn - Thanks for explaining. I think it's acceptable to implement this minor adjustment in the instrumentation library. However, it's important to note that the library cannot handle any other sequencing issues comprehensively. It primarily enhances activities according to the OTel specification. For any other requirements, users can utilize the `Enrich` API to customize existing activities.

**Comment by PascalSenn** [comment]
@vishweshbankwar 
Do you mean with 

> I think it's acceptable to implement this minor adjustment in the instrumentation library

the GRPC code which is already in place or the change in this PR?

**Comment by vishweshbankwar** [comment]
> @vishweshbankwar Do you mean with
> 
> > I think it's acceptable to implement this minor adjustment in the instrumentation library
> 
> the GRPC code which is already in place or the change in this PR?

I meant the change in this PR.

**Comment by PascalSenn** [comment]
@vishweshbankwar 

Very well, i would assume it's also what most frameworks with custom semantic conventions (like RPC) would want to do. Additional properties are also not a problem. They most important part is the name of the root activity.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs [hunk 1]
```diff
@@ -241,7 +241,13 @@ public void OnStopActivity(Activity activity, object payload)
                     context.GetEndpoint() as RouteEndpoint)?.RoutePattern.RawText;
             if (!string.IsNullOrEmpty(routePattern))
             {
-                activity.DisplayName = GetDisplayName(context.Request.Method, routePattern);
+                // only override the display name if it was not set by the user
```
Comment (vishweshbankwar) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:244]: @PascalSenn - Could you add a unit test for this? These tests are more focused on `http.route` tag value.

Comment (PascalSenn) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:244]: You mean instead of the one in RoutingTests?

Comment (vishweshbankwar) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:244]: ok to add this in addition to unit test. @alanwest - Do you have any preference?

**Comment by alanwest** [comment]
Before we move forward with this PR, there's a few things we should discuss.

I captured some general thoughts/questions on #5417 which may be best to discuss in our next SIG meeting.

With respect specifically to this AspNetCore instrumentation, if we move forward with this PR, we should confirm how and if this can be solved in the event that ASP.NET 9 introduces native trace instrumentation (see https://github.com/dotnet/aspnetcore/issues/52439)

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs (hunk 1 above)
Comment (alanwest) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:244]: I think it makes sense to test this new behavior separate from the tests for http.route.

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/RouteTests/RoutingTests.cs [hunk 2]
```diff
@@ -85,9 +85,12 @@ public async Task TestHttpRoute(RoutingTestCases.TestCase testCase)
 
         // Activity.DisplayName should be a combination of http.method + http.route attributes, see:
         // https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-spans.md#name
-        var expectedActivityDisplayName = string.IsNullOrEmpty(expectedHttpRoute)
-            ? testCase.HttpMethod
-            : $"{testCase.HttpMethod} {expectedHttpRoute}";
+        // unless the user has overriden the route name
```
Comment (OptimusPi) [inline test/OpenTelemetry.Instrumentation.AspNetCore.Tests/RouteTests/RoutingTests.cs:88]: replace typo `overriden` with `overridden` to satisfy ms-misspell PR check.

**Review by OptimusPi — CHANGES_REQUESTED** [review]
Fix typo here to satisfy PR misspell bot.

**Comment by vishweshbankwar** [comment]
@PascalSenn - We discussed this during yesterday's SIG meeting. Before proceeding with committing this change, we'd like to wait and gather more information on how https://github.com/dotnet/aspnetcore/issues/52439 will address these scenarios. This is to ensure that we don't encounter similar issues when users upgrade to a new .NET version where instrumentation is natively implemented within ASP.NET Core.

**Comment by vishweshbankwar** [comment]
Closing this for now. See https://github.com/open-telemetry/opentelemetry-dotnet/pull/5402#issuecomment-1995982094

**Pull request closed without merging**